lib/ThreeElementVectorLoweringPass.cpp

```diff
@@ -294,24 +275,61 @@ bool vec3ShouldBeLowered(Module &M) {
   }
 }
 
-} // namespace
-
-PreservedAnalyses
-clspv::ThreeElementVectorLoweringPass::run(Module &M, ModuleAnalysisManager &) {
-  PreservedAnalyses PA;
-  if (!vec3ShouldBeLowered(M))
-    return PA;
-
-  runOnGlobals(M);
-  for (auto &F : M.functions()) {
-    BitcastUtils::RemovedCstExprFromFunction(&F);
-    runOnFunction(F);
+bool clspv::ThreeElementVectorLoweringPass::vec3BitcastInFunction(Function &F) {
+  // TODO(#816): remove this loop after final transition.
+  for (Instruction &I : instructions(F)) {
+    if (auto *Inst = dyn_cast<CastInst>(&I)) {
+      auto *Type = Inst->getSrcTy();
+      if (Type->isPointerTy() && !Type->isOpaquePointerTy()) {
+        auto *PointeeType = Type->getNonOpaquePointerElementType();
+        if (auto *VectorType = dyn_cast<FixedVectorType>(PointeeType)) {
+          if (VectorType->getElementCount().getKnownMinValue() == 3)
+            return true;
+        }
+      }
+    }
   }
 
-  cleanDeadFunctions();
-  cleanDeadGlobals();
-
-  return PA;
+  // This method only works for opaque pointers as it will fail with non opaque
+  // pointer to pointer.
+  for (Instruction &I : instructions(F)) {
+    // If we have a store instruction...
+    if (auto *store = dyn_cast<StoreInst>(&I)) {
```

Comment:
What happens with the following testcase:
```
define void @test(ptr addrspace(1) %data) {
entry:
  %gep1 = getelementptr <3 x i32>, ptr addrspace(1) %data, i32 0, i32 2
  %gep2 = getelementptr i32, ptr addrspace(1) %gep1, i32 3
  store i32 0, ptr addrspace(1) %gep2
  ret void
}
```

It seems like this code wouldn't identify this case as problematic, but clspv would want to infer the argument type as `<3 x i32>`. To be safe I think the code needs to handle more cases where  pointer type is inferred. I would think a load case would also be problematic:
```
define i32 @test(ptr addrspace(1) %data) {
entry:
  %gep1 = getelementptr <3 x i32>, ptr addrspace(1) %data, i32 0, i32 2
  %gep2 = getelementptr i32, ptr addrspace(1) %gep1, i32 3
  %ld = load i32, ptr addrspace(1) %gep2
  ret i32 %ld
}
```

Both examples access the "4th" element of three-element vector.

Reply:
I tried to generalize it more by detecting if there is an arg or instruction that was inferred of a type and used as another type. This should handle these cases and the previous store case. store instruction inside `InferType` function changes will look throw the store operands to infer the type if it have only pointer operands so no need to handle it explicitly.